distribution/packages/build.gradle

```diff
@@ -203,37 +146,166 @@ subprojects {
+  }
+
+  requires('coreutils')
+
+  fileMode 0644
+  dirMode 0755
+  user 'root'
+  permissionGroup 'root'
+
+  into '/usr/share/elasticsearch'
+  with libFiles
+  with modulesFiles
+  with noticeFile
+}
+
+task buildDeb(type: Deb) {
+  dependsOn processDebFiles
+  configure(commonPackageConfig('deb'))
+
+  version = project.version
+  packageGroup 'web'
+  requires 'bash'
+  requires 'libc6'
+  requires 'adduser'
+
+  into('/usr/share/lintian/overrides') {
+    from('src/deb/lintian/elasticsearch')
+  }
+  into('/usr/share/doc/elasticsearch') {
+    from 'src/deb/copyright'
+    fileMode 0644
+  }
+}
+
+// task that sanity checks if the Deb archive can be extracted
+task checkDeb(type: LoggedExec) {
+  onlyIf { new File('/usr/bin/dpkg-deb').exists() || new File('/usr/local/bin/dpkg-deb').exists() }
+  final File debExtracted = new File("${buildDir}", 'deb-extracted')
+  commandLine 'dpkg-deb', '-x', "${buildDir}/distributions/elasticsearch-${project.version}.deb", debExtracted
+  doFirst {
+    debExtracted.deleteDir()
+  }
+}
+
+task buildRpm(type: Rpm) {
+  dependsOn processRpmFiles
+  configure(commonPackageConfig('rpm'))
+
+  packageGroup 'Application/Internet'
+  requires '/bin/bash'
+
+  prefix '/usr'
+  packager 'Elasticsearch'
+  version = project.version.replace('-', '_')
+  release = '1'
+  arch 'NOARCH'
+  os 'LINUX'
+  license '2009'
+  distribution 'Elasticsearch'
+  vendor 'Elasticsearch'
+  // TODO ospackage doesn't support icon but we used to have one
+
+  // without this the rpm will have parent dirs of any files we copy in, eg /etc/elasticsearch
+  addParentDirs false
+
+  // Declare the folders so that the RPM package manager removes
+  // them when upgrading or removing the package
+  directory('/usr/share/elasticsearch/bin', 0755)
+  directory('/usr/share/elasticsearch/lib', 0755)
+  directory('/usr/share/elasticsearch/modules', 0755)
+  modulesFiles.eachFile { FileCopyDetails fcp ->
+    if (fcp.name == "plugin-descriptor.properties") {
+      directory('/usr/share/elasticsearch/modules/' + fcp.file.parentFile.name, 0755)
+    }
   }
 }
 
+// task that sanity checks if the RPM archive can be extracted
+task checkRpm(type: LoggedExec) {
+  onlyIf { new File('/bin/rpm').exists() || new File('/usr/bin/rpm').exists() || new File('/usr/local/bin/rpm').exists() }
+  final File rpmDatabase = new File("${buildDir}", 'rpm-database')
+  final File rpmExtracted = new File("${buildDir}", 'rpm-extracted')
+  commandLine 'rpm',
+          '--badreloc',
+          '--nodeps',
+          '--noscripts',
+          '--notriggers',
+          '--dbpath',
+          rpmDatabase,
+          '--relocate',
+          "/=${rpmExtracted}",
+          '-i',
+          "${buildDir}/distributions/elasticsearch-${project.version}.rpm"
+  doFirst {
+    rpmDatabase.deleteDir()
+    rpmExtracted.deleteDir()
+  }
+}
+
+subprojects {
+  apply plugin: 'distribution'
```

Comment:
I think it is worth explaining here that we keep the subprojects at all so we can use project substitution to resolve to the individual artifacts because project substitution only works with the default configuration.